### ISSUE TITLE
#js triggers "no reader function for tag js"

### ISSUE THREAD

**Issue by sritchie** [nextjournal/clerk#249]
You can reproduce this in any clerk notebook with the following form:

```clj
(clerk/with-viewer
  '(fn [_]
     (v/html
      [:div "error: " (str #js [1 2 3])]))
  "")
```

error:

<img width="927" alt="image" src="https://user-images.githubusercontent.com/69635/198755361-9856886a-ab54-476d-9c6d-8185d2dacf7a.png">

cc @borkdude

**Comment by borkdude** [comment]
This is because a quoted expressions is read in the host (JVM Clojure) first. There is no way around this unless you use a string or write `js-obj` instead. You can also use `eval-clj-str` in Clerk.

**Comment by borkdude** [comment]
Another way around this would be to handle the `#js` tag on the Clojure side as well. cc @mk 

In JVM Clojure:
``` clojure
user=> (e/parse-string "#js {:a 1}" {:readers {'js (fn [data] (if (vector? data) (list* 'cljs.core/array data) (list* 'cljs.core/js-obj (let [ks (keys data) ks (map name ks) vs (vals data)] (interleave ks vs)))))}})
(cljs.core/js-obj "a" 1)

user=> (e/parse-string "#js [1 2 3]" {:readers {'js (fn [data] (if (vector? data) (list* 'cljs.core/array data) (list* 'cljs.core/js-obj (let [ks (keys data) ks (map name ks) vs (vals data)] (interleave ks vs)))))}})
(cljs.core/array 1 2 3)
```

**Comment by sritchie** [comment]
👍 of course that makes sense. Thanks @borkdude !

**Comment by mk** [comment]
For `#js [,,,]` the rewrite is trivial:

> The tagged literal `#js [1 2 3]` is equivalent to `(array 1 2 3)`

For maps the rewriting they keys is more tedious. It's also another thing users need to know about.

Since Clerk is doing it's own read we could easily support `#js` there, though that comes with the downside of not being the same at the REPL anymore. Otoh setting up a global reader for `#js` also feels a bit invasive?

**Comment by borkdude** [comment]
How is it more tedious than what I showed above?

**Comment by mk** [comment]
@borkdude I mean it's tedious for a user to convert an existing js object literal `#js {:hello "world" :another "kv"}` to `(js-obj "hello" "world" "another" "kv")`. What you've shown above is also tedious. This only applies for when we don't add `#js` support of course.

**Comment by borkdude** [comment]
What I've shown above is how you can add the above to clerk's usage of edamame to support it.

**Comment by borkdude** [comment]
Right. You could expose that function so people can add it to data readers themselves:

```
(set! *data-readers* (assoc *data-readers* 'js clerk/js-literal))
```

**Comment by mk** [comment]
<img width="864" alt="Screen Shot 2023-02-10 at 11 42 15" src="https://user-images.githubusercontent.com/1187/218072223-4f3533e1-cd8d-4137-bf78-98346248c605.png">

**Comment by sritchie** [comment]
Looking great!!

**Comment by borkdude** [comment]
Nice, but a pity I didn't get credits / co-author in the commit while I basically wrote the solution? :P

**Comment by mk** [comment]
@borkdude fixed with 477c2c3.

**Comment by borkdude** [comment]
Yaaay, thank you hehee : )
